native/lib/ruby_driver/node_converter.rb

```diff
@@ -0,0 +1,259 @@
+require 'parser/current'
+
+# Nodes doc:
+# https://github.com/whitequark/parser/blob/master/doc/AST_FORMAT.md
+
+module NodeConverter
+  class Converter
+    def initialize(node, comments)
+      if not node.is_a?(Parser::AST::Node)
+        raise "Object is not a Parser::AST::Node"
+      end
+
+      @root = node
+      @comments = comments
+      @dict = {}
+
+    end
+
+    def tohash()
+      @dict["ast"] = {}
+      @dict["ast"]["module"] = convert(@root)
+      add_comments()
+      return @dict["ast"]
+    end
+
+    private
+
+    def convert(node)
```

Comment:
This looks pretty complex and hard to maintain...
Can you elaborate a bit on what kind of problem are you trying to solve here? It looks as if some custom logic is being built on top of the native AST, instead of just doing a generic S-expression to JSON conversion.

Reply:
The values of the s-expressions have meanings depending on their positions inside the lists. This is really hard to debug when checking a complex native AST since you've to check the docs for every single node you're seeing to check what every position means so instead of converting to a (json-exportable) array, I convert them to a keyed dictionary, which is somewhat like incorporating the documentation inside the node (and also will make it easier to write the annotations later).

Actually I started doing this as a debugging/checking aid and then just tough that since it was done it could stay that way.

Also, some nodes have more complex logic to get the values like the token and such; this surely could be done in the annotations DSL but ruby is just more expressive/easier.

So I'm basically trading a little more maintain load in the native driver for easier debugging of problems in the native UAST and a simpler annotations code.

Reply:
Also, it must be a dictionary anyway since the positions in the native objects are not in the s-exps but on a custom object accessed from the "loc" method (which can't be directly exported to JSON as such).

Reply:
I see why you may want to convert from sexp to a dict, but I fail to see why it can't be done in a generic way and every keyword is handled differently instead, could you include some example of a sexp that exposes the problem?

Reply:
The problem is that there aren't keywords (what this does is adding them), just positions inside the list with different meanings depending on the type. The API doesn't have any way to do a query like "for this sexp, the first element is the constant base and the second the value".

See: https://github.com/whitequark/parser/blob/master/doc/AST_FORMAT.md

Reply:
Thanks for the link, I understand it now.